### PR TITLE
Add a stone on board

### PR DESCRIPTION
**Pull request by Chroq**
*Trello* : https://trello.com/c/35WRWcSc/2-05-as-norbert-i-want-to-be-able-to-add-a-stone-anywhere-on-the-board

In this PR, we can put a stone on the board.

- If the stone is outside the board, you're notified and you can retry.
- if a stone already exists on the selected coordinate, you're notified and you can retry.
- Incorrect inputs are handled 

## Expected results

*make start and play some stones*
Capture d’écran de 2019-11-20 12-52-33

*make test*
Capture d’écran de 2019-11-20 12-52-01

### CODE REVIEW

File: src/board/board.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    lenght = len(pattern)
+
+    if is_correct_position(lenght, line, column):
+        raise ValueError("Stone can't be put outside the board")
+
+    if is_already_taken_place(pattern, line, column):
+        raise ValueError("A stone already exists at this position")
+
+    pattern[column][line] = 1
```
Comment (zyhou) [inline src/board/board.py]: Attention, `pattern` is mutate in this case, can you clone this param?

File: src/board/board.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+from board.coord import get_x_index
+
+
+def put_stone(coords, pattern):
+    """ Put a stone on the coords provided."""
```
Comment (zyhou) [inline src/board/board.py]: For me, the comment aren't necessary. The function and the args already explain well what he does

File: src/board/board.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    if coords is None:
+        raise ValueError("Coordinates for the stone is incorrect. Please, pick a new one")
+
+    line = get_x_index(coords[0]) - 1
```
Comment (zyhou) [inline src/board/board.py]: no need to convert to int? Or why the above line you use the `int` function ?

File: src/board/board.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    pattern[column][line] = 1
+
+    return pattern
+
+
+def is_correct_position(length, line, column):
+    """ Check if the stone can be put inside the board """
+
+    return length > line & length > column
```
Comment (zyhou) [inline src/board/board.py]: uise `and` operator (@Kmaschta it's the same or not ? don't remember..)

File: src/board/coord.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+def get_x_index(letter):
+    """ The X index is the "line" part of the board. Because coordinate is a letter to represent her,
+    we need his position in alphabet.
+    To begin at the 1 index, we add 1 to list index. """
+    return list(string.ascii_uppercase).index(letter.upper()) + 1
+
+
+def get_coord(position):
+    """ Using a regex, this function checks if the position is correct """
+    regex = re.match("^([a-zA-Z]{1})([0-9]{1,2})$", position)
```
Comment (zyhou) [inline src/board/coord.py]: Not sure about this one. Come to see me :)

File: src/board/test_board.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+from board.board import put_stone, is_correct_position, is_already_taken_place
+
+
+def test_can_put_a_stone_on_board():
```
Comment (zyhou) [inline src/board/test_board.py]: All you test is well name :+1:

File: src/main.py [hunk 7]
```diff
@@ -1,11 +1,31 @@
...
+pattern = generate(9)
 output = render(pattern)
 
-print(output)
+still_in_game = True
+print(render(pattern))
+
+while still_in_game:
+
+    position = input("Stone on : ")
```
Comment (zyhou) [inline src/main.py]: what append if I enter another char ? (non valid char)

File: src/board/board.py (hunk 4 above)
Comment (Chroq) [reply src/board/board.py]: Not the same, indeed ! I changed for _and_ operator in the last commit :)

File: src/main.py (hunk 7 above)
Comment (Chroq) [reply src/main.py]: We get back a string from _input_ function and this string is validated by the regex right after

File: src/board/board.py (hunk 1 above)
Comment (Chroq) [reply src/board/board.py]: Changed ! TY

File: src/board/board.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+    length = len(board)
+
+    if is_outside(length, line, column):
+        raise ValueError("Stone can't be put outside the board")
+
+    if is_already_taken_place(board, line, column):
+        raise ValueError("A stone already exists at this position")
+
+    new_board[column - 1][line - 1] = 1
```
Comment (zyhou) [inline src/board/board.py]: :+1: and better name

**Comment by alexisjanvier** [comment]
The column names and row numbers are missing in the board display, no? something like : 
```
  A B C D E
1 ⬡ ⬡ ⬡ ⬡ ⬡ 
2   ⬡ ⬡ ⬡ ⬡ ⬡ 
3     ⬡ ⬡ ⬡ ⬡ ⬡ 
```
(without this, it is not easy to indicate a coordinate ...)

File: src/board/generator.py [hunk 9]
```diff
@@ -1,8 +1,8 @@
-
 def generate(size):
     """"
     Returns an empty board
     :parameter size int Size board. Classical size are 9, 11, 14 or 19
     :return board array Array representation of board
     """
-    return [[ 0 for i in range(size)] for j in range(size) ]
+    # return [[0] * size] * size
```
Comment (Luwangel) [inline src/board/generator.py]: Comment spotted :eyes:

File: src/board/coord.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import string
```
Comment (Luwangel) [inline src/board/coord.py]: I think the notion of coordinates should be in the IO folder. The board is represented as a 2D array of integers. And so the letters are used to the end-user, and not the model.

**Pull request merged**